File: packages/shared/src/service-auth.ts
/**
 * Per-service request authentication (the `sig1` signature format).
 *
 * Replaces the shared claimless bearer for service-to-service traffic: each
 * service signs its requests with its own secret, and the signature binds the
 * full request (method, path, query, body hash, asserted actor) so a captured
 * credential cannot be replayed against a different request.
 *
 * The canonical request string layout is a cross-language contract with
 * `sandbox_runtime/auth/service_auth.py`, pinned by the golden vectors in
 * `test-fixtures/service-auth-vectors.json`. Any change to the layout or the
 * canonicalization rules requires a new format tag (`sig2`), not an edit here.
 */

import { bytesToHex, computeHmacHex, timingSafeEqual, TOKEN_VALIDITY_MS } from "./auth";

export const SERVICE_HEADER = "X-OpenInspect-Service";
export const SERVICE_SIGNATURE_HEADER = "X-OpenInspect-Service-Signature";
export const ACTOR_HEADER = "X-OpenInspect-Actor";
export const SIG1_PREFIX = "sig1";

export const SERVICE_NAMES = ["web", "slack-bot", "github-bot", "linear-bot", "modal"] as const;
export type ServiceName = (typeof SERVICE_NAMES)[number];

export function isServiceName(value: string): value is ServiceName {
  return (SERVICE_NAMES as readonly string[]).includes(value);
}

/** Signature verification failure, ordered from cheapest to most specific check. */
export type ServiceSignatureFailure = "format" | "expired" | "mismatch";

/**
 * Successful verification returns the parsed wire components so callers never
 * re-split the header (this module is the sole owner of the sig1 grammar).
 */
export type ServiceSignatureResult =
  | { ok: true; timestampMs: number; nonce: string }
  | { ok: false; reason: ServiceSignatureFailure };

const NONCE_PATTERN = /^[0-9a-f]{1,64}$/;
// Strict ASCII decimal, mirrored by service_auth.py. Number()'s wider grammar
// ("1e3", "0x10", padding) must not classify differently across languages.
const TIMESTAMP_PATTERN = /^[0-9]{1,16}$/;

/**
 * SHA-256 of the raw request body as lowercase hex. Bodyless requests hash
 * the empty byte string.
 */
export async function sha256Hex(data: ArrayBuffer | Uint8Array | string): Promise<string> {
  const bytes =
    typeof data === "string"
      ? new TextEncoder().encode(data)
      : data instanceof Uint8Array
        ? data
        : new Uint8Array(data);
  const digest = await crypto.subtle.digest("SHA-256", bytes);
  return bytesToHex(new Uint8Array(digest));
}

function compareUtf8Bytes(a: string, b: string): number {
  const encoder = new TextEncoder();
  const aBytes = encoder.encode(a);
  const bBytes = encoder.encode(b);
  const len = Math.min(aBytes.length, bBytes.length);
  for (let i = 0; i < len; i++) {
    if (aBytes[i] !== bBytes[i]) {
      return aBytes[i] - bBytes[i];
    }
  }
  return aBytes.length - bBytes.length;
}

/**
 * Canonical form of a URL's query string: decoded `key=value` entries sorted
 * bytewise (UTF-8) by `key\0value`, re-encoded with `encodeURIComponent`, and
 * joined with `&`. An empty query canonicalizes to the empty string, so
 * `?a=1&b=2` and `?b=2&a=1` sign identically while `?a=1` and no query do not
 * collide with each other's canonical strings.
 */
export function canonicalizeQuery(search: string): string {
  const params = new URLSearchParams(search);
  const entries = Array.from(params.entries());
  entries.sort((a, b) => compareUtf8Bytes(`${a[0]}\0${a[1]}`, `${b[0]}\0${b[1]}`));
  return entries
    .map(([key, value]) => `${encodeURIComponent(key)}=${encodeURIComponent(value)}`)
    .join("&");
}

/**
 * The exact byte layout signed by `sig1`. Every field is newline-delimited;
 * `actor` is the empty string when no actor header is sent.
 */
export function buildCanonicalRequestString(p: {
  service: ServiceName;
  timestampMs: number;
  nonce: string;
  method: string;
  pathname: string;
  canonicalQuery: string;
  bodySha256Hex: string;
  actor: string;
}): string {
  return (
    `${SIG1_PREFIX}\n${p.service}\n${p.timestampMs}\n${p.nonce}\n` +
    `${p.method.toUpperCase()}\n${p.pathname}\n${p.canonicalQuery}\n` +
    `${p.bodySha256Hex}\n${p.actor}`
  );
}

function generateNonce(): string {
  const bytes = new Uint8Array(8);
  crypto.getRandomValues(bytes);
  return bytesToHex(bytes);
}

async function signCanonicalRequest(p: {
  service: ServiceName;
  secret: string;
  timestampMs: number;
  nonce: string;
  method: string;
  url: string;
  bodySha256Hex: string;
  actor: string;
}): Promise<string> {
  const parsed = new URL(p.url);
  const canonical = buildCanonicalRequestString({
    service: p.service,
    timestampMs: p.timestampMs,
    nonce: p.nonce,
    method: p.method,
    pathname: parsed.pathname,
    canonicalQuery: canonicalizeQuery(parsed.search),
    bodySha256Hex: p.bodySha256Hex,
    actor: p.actor,
  });
  return computeHmacHex(canonical, p.secret);
}

/**
 * Build the sig1 request headers for an outbound service call.
 *
 * Returns `X-OpenInspect-Service`, `X-OpenInspect-Service-Signature`, plus
 * `X-OpenInspect-Actor` when an actor is asserted and `x-trace-id` when a
 * trace ID is provided. Callers add their own `Content-Type`/`Accept`
 * headers, and must send exactly the body bytes that were signed.
 */
export async function buildServiceAuthHeaders(p: {
  service: ServiceName;
  secret: string;
  method: string;
  url: string;
  body?: ArrayBuffer | Uint8Array | string;
  actor?: string;
  traceId?: string;
}): Promise<Record<string, string>> {
  const timestampMs = Date.now();
  const nonce = generateNonce();
  const actor = p.actor ?? "";
  const bodySha256Hex = await sha256Hex(p.body ?? "");
  const signature = await signCanonicalRequest({
    service: p.service,
    secret: p.secret,
    timestampMs,
    nonce,
    method: p.method,
    url: p.url,
    bodySha256Hex,
    actor,
  });

  const headers: Record<string, string> = {
    [SERVICE_HEADER]: p.service,
    [SERVICE_SIGNATURE_HEADER]: `${SIG1_PREFIX}.${timestampMs}.${nonce}.${signature}`,
  };
  if (actor) {
    headers[ACTOR_HEADER] = actor;
  }
  if (p.traceId) {
    headers["x-trace-id"] = p.traceId;
  }
  return headers;
}

export type ServiceSignatureHeaderParse =
  | { ok: true; timestampMs: number; nonce: string; signature: string }
  | { ok: false; reason: "format" | "expired" };

/**
 * Parse a sig1 signature header and check everything that does not need the
 * request body: grammar, then timestamp freshness. Verifiers call this before
 * buffering the body so a malformed or stale header never costs a body read —
 * only the HMAC comparison itself needs the body hash.
 */
export function parseServiceSignatureHeader(header: string): ServiceSignatureHeaderParse {
  const parts = header.split(".");
  if (parts.length !== 4 || parts[0] !== SIG1_PREFIX) {
    return { ok: false, reason: "format" };
  }
  const [, timestampPart, nonce, signature] = parts;
  if (!TIMESTAMP_PATTERN.test(timestampPart)) {
    return { ok: false, reason: "format" };
  }
  const timestampMs = Number(timestampPart);
  if (timestampMs <= 0) {
    return { ok: false, reason: "format" };
  }
  if (!NONCE_PATTERN.test(nonce) || !/^[0-9a-f]{64}$/.test(signature)) {
    return { ok: false, reason: "format" };
  }
  if (Math.abs(Date.now() - timestampMs) > TOKEN_VALIDITY_MS) {
    return { ok: false, reason: "expired" };
  }
  return { ok: true, timestampMs, nonce, signature };
}

/**
 * Verify a sig1 signature header against the named service's secret and the
 * request it claims to sign. The caller supplies the body hash (the body must
 * be buffered exactly once at the edge) and the actor header value (empty
 * string when absent), both of which are covered by the signature.
 */
export async function verifyServiceSignature(p: {
  signatureHeader: string;
  service: ServiceName;
  secret: string;
  method: string;
  url: string;
  bodySha256Hex: string;
  actor: string;
}): Promise<ServiceSignatureResult> {
  const parsed = parseServiceSignatureHeader(p.signatureHeader);
  if (!parsed.ok) {
    return parsed;
  }
  const expected = await signCanonicalRequest({
    service: p.service,
    secret: p.secret,
    timestampMs: parsed.timestampMs,
    nonce: parsed.nonce,
    method: p.method,
    url: p.url,
    bodySha256Hex: p.bodySha256Hex,
    actor: p.actor,
  });
  if (!timingSafeEqual(parsed.signature, expected)) {
    return { ok: false, reason: "mismatch" };
  }
  return { ok: true, timestampMs: parsed.timestampMs, nonce: parsed.nonce };
}

/** A sender's outbound credential: sig1 with the sender's own secret. */
export interface OutboundServiceCredential {
  service: ServiceName;
  secret: string;
}

/** The env surface every sender carries. */
export interface OutboundCredentialEnv {
  SERVICE_AUTH_SECRET?: string;
}

export function resolveOutboundCredential(
  service: ServiceName,
  env: OutboundCredentialEnv
): OutboundServiceCredential {
  if (!env.SERVICE_AUTH_SECRET) {
    throw new Error(`SERVICE_AUTH_SECRET is required for outbound ${service} requests`);
  }
  return { service, secret: env.SERVICE_AUTH_SECRET };
}

/**
 * A non-JSON outbound body: the exact serialized bytes plus the content type
 * describing them (e.g. a buffered multipart form with its boundary). Binary
 * bodies must be serialized BEFORE signing — sig1 hashes the exact bytes sent.
 */
export interface OutboundBinaryBody {
  bytes: Uint8Array;
  contentType: string;
}

/** One outbound control-plane request to authenticate. */
export interface OutboundRequestToSign {
  method: string;
  url: string;
  /** The exact body being sent, when there is one: a JSON string, or pre-serialized bytes with their content type. */
  body?: string | OutboundBinaryBody;
  /** External actor this request acts for (`slack:U…`), asserted under the sig1 signature. */
  actor?: string;
  traceId?: string;
}

/** The BodyInit to send for a request's signed body — exactly what was signed. */
function outboundBodyInit(body: OutboundRequestToSign["body"]): string | Uint8Array | undefined {
  return typeof body === "object" ? body.bytes : body;
}

/**
 * Build the auth headers for an outbound control-plane request from the
 * resolved credential. Signatures bind method, URL, body, and actor, so
 * headers are built per request — never shared across calls — and callers
 * must send exactly the body they signed. A `Content-Type` header is included
 * only when a body is present: `application/json` for string bodies, the
 * body's own content type for binary bodies.
 */
export async function buildOutboundAuthHeaders(
  credential: OutboundServiceCredential,
  request: OutboundRequestToSign
): Promise<Record<string, string>> {
  const contentType: Record<string, string> =
    request.body === undefined
      ? {}
      : {
          "Content-Type":
            typeof request.body === "string" ? "application/json" : request.body.contentType,
        };
  return {
    ...contentType,
    ...(await buildServiceAuthHeaders({
      service: credential.service,
      secret: credential.secret,
      method: request.method,
      url: request.url,
      body: outboundBodyInit(request.body),
      actor: request.actor,
      traceId: request.traceId,
    })),
  };
}

/**
 * Minimal interface for the control-plane service binding. Compatible with
 * Cloudflare Workers' `Fetcher` type without depending on
 * `@cloudflare/workers-types`.
 */
export interface ControlPlaneFetcher {
  fetch(input: string | URL | Request, init?: RequestInit): Promise<Response>;
}

/** Options a signed control-plane fetch accepts beyond the request being signed. */
export interface SignedFetchInit {
  /** Non-auth headers (e.g. `Accept`); the auth headers always take precedence. */
  headers?: Record<string, string>;
  signal?: AbortSignal;
}

/**
 * Sign one outbound control-plane request as `service` and send it through
 * the env's `CONTROL_PLANE` service binding. Signing and sending consume the
 * same method/URL/body values, so the bytes sent are exactly the bytes signed
 * by construction — callers never handle the signed body themselves.
 */
export async function signedControlPlaneFetch(
  service: ServiceName,
  env: OutboundCredentialEnv & { CONTROL_PLANE: ControlPlaneFetcher },
  request: OutboundRequestToSign,
  init?: SignedFetchInit
): Promise<Response> {
  const credential = resolveOutboundCredential(service, env);
  const headers = { ...init?.headers, ...(await buildOutboundAuthHeaders(credential, request)) };
  return env.CONTROL_PLANE.fetch(request.url, {
    method: request.method,
    headers,
    body: outboundBodyInit(request.body),
    signal: init?.signal,
  });
}

File: packages/sandbox-runtime/src/sandbox_runtime/auth/service_auth.py
"""Per-service request authentication (the ``sig1`` signature format).

Python mirror of ``packages/shared/src/service-auth.ts``. The canonical
request string layout is a cross-language contract pinned by the golden
vectors in ``packages/shared/test-fixtures/service-auth-vectors.json``; any
change to the layout or the canonicalization rules requires a new format tag
(``sig2``), not an edit here.

Internals use seconds per repo convention; the wire string carries epoch
milliseconds to match the TypeScript side.
"""

import hashlib
import hmac
import re
import secrets
import time
from typing import Literal, NamedTuple
from urllib.parse import parse_qsl, quote, urlsplit

from .internal import TOKEN_VALIDITY_SECONDS

SERVICE_HEADER = "X-OpenInspect-Service"
SERVICE_SIGNATURE_HEADER = "X-OpenInspect-Service-Signature"
ACTOR_HEADER = "X-OpenInspect-Actor"
SIG1_PREFIX = "sig1"

# Characters encodeURIComponent leaves unescaped, beyond Python quote()'s
# always-safe alphanumerics and "_.-~".
_ENCODE_URI_COMPONENT_SAFE = "!'()*"

# WHATWG URL parsers (the TypeScript signer and the Workers runtime) keep
# these path characters literal while percent-encoding non-ASCII bytes.
_PATH_SAFE = "/:@!$&'()*+,;=-._~%[]"

# Path characters (and dot segments) where quote() and WHATWG serialization
# are known to diverge (e.g. "\\" becomes "/" in WHATWG special URLs). The
# signer refuses them loudly instead of producing a signature the TypeScript
# verifier can never match.
_PATH_UNVETTED_CHARS = set('\\|^`<>"{} ')

_HEX_NONCE_MAX_LEN = 64

# Strict ASCII decimal, mirrored by service-auth.ts. str.isdigit()'s wider
# grammar (non-ASCII Unicode digits) must not classify differently across
# languages.
_TIMESTAMP_PATTERN = re.compile(r"[0-9]{1,16}")

ServiceSignatureFailure = Literal["format", "expired", "mismatch"]


class ServiceSignatureResult(NamedTuple):
    """Outcome of ``verify_service_signature``.

    On success, carries the parsed wire components so callers never re-split
    the header (this module is the sole owner of the sig1 grammar).
    """

    ok: bool
    reason: ServiceSignatureFailure | None
    timestamp_ms: int | None = None
    nonce: str | None = None


def sha256_hex(data: bytes | str) -> str:
    """SHA-256 of the raw request body as lowercase hex ("" for no body)."""
    raw = data.encode("utf-8") if isinstance(data, str) else data
    return hashlib.sha256(raw).hexdigest()


def canonicalize_query(query: str) -> str:
    """Canonical form of a URL query string.

    Decoded ``key=value`` entries sorted bytewise (UTF-8) by ``key\\0value``,
    re-encoded with encodeURIComponent semantics, joined with ``&``.
    """
    entries = parse_qsl(query.lstrip("?"), keep_blank_values=True)
    entries.sort(key=lambda kv: f"{kv[0]}\0{kv[1]}".encode())
    return "&".join(
        f"{quote(key, safe=_ENCODE_URI_COMPONENT_SAFE)}"
        f"={quote(value, safe=_ENCODE_URI_COMPONENT_SAFE)}"
        for key, value in entries
    )


def _canonical_pathname(url: str) -> str:
    """The URL path as a WHATWG parser would serialize it.

    Already-encoded paths pass through unchanged (``%`` is safe); raw
    non-ASCII input is percent-encoded to match ``new URL(url).pathname``.
    """
    return quote(urlsplit(url).path, safe=_PATH_SAFE)


def _validate_signable_path(url: str) -> None:
    """Refuse to sign paths whose WHATWG serialization we do not mirror.

    ``_canonical_pathname`` approximates WHATWG for the vetted character set;
    outside it (backslash, ``|``, ``^``, dot segments, controls) the two
    serializations diverge and the signature would fail verification as an
    opaque 401 far from the cause. Failing loudly at the signer keeps the
    contract honest without emulating the full WHATWG algorithm.
    """
    path = urlsplit(url).path
    for ch in path:
        if ch in _PATH_UNVETTED_CHARS or ord(ch) < 0x20 or ch == "\x7f":
            raise ValueError(
                f"cannot sign path containing {ch!r}: its WHATWG serialization "
                "is not mirrored here; percent-encode it before signing"
            )
    if any(segment in (".", "..") for segment in path.split("/")):
        raise ValueError(
            "cannot sign path containing dot segments: WHATWG parsers resolve "
            "them; resolve the path before signing"
        )


def build_canonical_request_string(
    *,
    service: str,
    timestamp_ms: int,
    nonce: str,
    method: str,
    pathname: str,
    canonical_query: str,
    body_sha256_hex: str,
    actor: str,
) -> str:
    """The exact byte layout signed by ``sig1`` (actor is "" when absent)."""
    return (
        f"{SIG1_PREFIX}\n{service}\n{timestamp_ms}\n{nonce}\n"
        f"{method.upper()}\n{pathname}\n{canonical_query}\n"
        f"{body_sha256_hex}\n{actor}"
    )


def _sign_canonical_request(
    *,
    service: str,
    secret: str,
    timestamp_ms: int,
    nonce: str,
    method: str,
    url: str,
    body_sha256_hex: str,
    actor: str,
) -> str:
    canonical = build_canonical_request_string(
        service=service,
        timestamp_ms=timestamp_ms,
        nonce=nonce,
        method=method,
        pathname=_canonical_pathname(url),
        canonical_query=canonicalize_query(urlsplit(url).query),
        body_sha256_hex=body_sha256_hex,
        actor=actor,
    )
    return hmac.new(
        secret.encode("utf-8"),
        canonical.encode("utf-8"),
        hashlib.sha256,
    ).hexdigest()


def build_service_auth_headers(
    *,
    service: str,
    secret: str,
    method: str,
    url: str,
    body: bytes | str | None = None,
    actor: str | None = None,
    trace_id: str | None = None,
) -> dict[str, str]:
    """Build the sig1 request headers for an outbound service call.

    Callers add their own ``Content-Type``/``Accept`` headers and must send
    exactly the body bytes that were signed.
    """
    _validate_signable_path(url)
    timestamp_ms = int(time.time() * 1000)
    nonce = secrets.token_hex(8)
    actor_value = actor or ""
    body_sha256_hex = sha256_hex(body if body is not None else b"")
    signature = _sign_canonical_request(
        service=service,
        secret=secret,
        timestamp_ms=timestamp_ms,
        nonce=nonce,
        method=method,
        url=url,
        body_sha256_hex=body_sha256_hex,
        actor=actor_value,
    )

    headers = {
        SERVICE_HEADER: service,
        SERVICE_SIGNATURE_HEADER: f"{SIG1_PREFIX}.{timestamp_ms}.{nonce}.{signature}",
    }
    if actor_value:
        headers[ACTOR_HEADER] = actor_value
    if trace_id:
        headers["x-trace-id"] = trace_id
    return headers


def _is_lower_hex(value: str, *, max_len: int) -> bool:
    return 0 < len(value) <= max_len and all(c in "0123456789abcdef" for c in value)


def verify_service_signature(
    *,
    signature_header: str,
    service: str,
    secret: str,
    method: str,
    url: str,
    body_sha256_hex: str,
    actor: str,
) -> ServiceSignatureResult:
    """Verify a sig1 signature header against the named service's secret."""
    parts = signature_header.split(".")
    if len(parts) != 4 or parts[0] != SIG1_PREFIX:
        return ServiceSignatureResult(ok=False, reason="format")
    _, timestamp_part, nonce, signature = parts
    if not _TIMESTAMP_PATTERN.fullmatch(timestamp_part):
        return ServiceSignatureResult(ok=False, reason="format")
    timestamp_ms = int(timestamp_part)
    if timestamp_ms <= 0:
        return ServiceSignatureResult(ok=False, reason="format")
    if not _is_lower_hex(nonce, max_len=_HEX_NONCE_MAX_LEN):
        return ServiceSignatureResult(ok=False, reason="format")
    if len(signature) != 64 or not _is_lower_hex(signature, max_len=64):
        return ServiceSignatureResult(ok=False, reason="format")
    if abs(time.time() - timestamp_ms / 1000) > TOKEN_VALIDITY_SECONDS:
        return ServiceSignatureResult(ok=False, reason="expired")
    expected = _sign_canonical_request(
        service=service,
        secret=secret,
        timestamp_ms=timestamp_ms,
        nonce=nonce,
        method=method,
        url=url,
        body_sha256_hex=body_sha256_hex,
        actor=actor,
    )
    if not hmac.compare_digest(signature, expected):
        return ServiceSignatureResult(ok=False, reason="mismatch")
    return ServiceSignatureResult(ok=True, reason=None, timestamp_ms=timestamp_ms, nonce=nonce)
